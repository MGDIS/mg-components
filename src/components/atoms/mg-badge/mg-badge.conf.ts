/**
 * List of all possibles variants
 */
export const variants = ['info', 'primary', 'secondary', 'success', 'warning', 'danger'] as const;

/**
 * Variant type from variants
 */
export type BadgeVariantType = (typeof variants)[number];
