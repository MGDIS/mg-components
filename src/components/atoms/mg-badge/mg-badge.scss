@mixin badge-variant($name, $backgroundColor, $color) {
  &.mg-badge--#{$name} {
    // Decoration
    background-color: hsl(var(--color-#{$backgroundColor}));
    border-color: hsl(var(--color-#{$backgroundColor}));
    // Font
    color: hsl(var(--color-#{$color}));

    &.mg-badge--outline {
      // Decoration
      background-color: transparent;
      // Font
      color: hsl(var(--color-#{$backgroundColor}));
    }
  }
}

.mg-badge {
  $vertical-padding: 0.3rem;
  $line-height: var(--mg-badge-font-size);
  // Display
  padding: $vertical-padding 0.7em;
  min-height: calc($vertical-padding * 2 + $line-height);
  box-sizing: border-box;
  // Decoration
  border-radius: var(--mg-badge-border-radius);
  border: 0.1rem solid;
  // Font
  font-size: var(--mg-badge-font-size);
  line-height: $line-height;
  font-weight: 600;
  text-align: center;

  // Variant
  @include badge-variant('primary', 'dark', 'light');
  @include badge-variant('secondary', 'light', 'dark');
  @include badge-variant('success', 'success', 'dark');
  @include badge-variant('warning', 'warning', 'dark');
  @include badge-variant('danger', 'danger', 'light');
  @include badge-variant('info', 'info', 'light');
}

.mg-badge,
mg-badge {
  display: inline-block;
  vertical-align: middle;
}
