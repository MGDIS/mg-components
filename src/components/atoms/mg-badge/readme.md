## Design

The badge is always placed on top of, or next to the element it is for.

The badge displays a number (can be followed by the `+` character) or a punctuation character.

## Specs

![](./mg-badge/doc/img/mg-badge-specs.png)

## Theming

![](./mg-badge/doc/img/mg-badge-styles.png)

## CSS Variables

If needed some [variables](./?path=/story/css-variables--page) are available to customize the component:

- `--mg-badge-size`: define badge height and min-width, default: `1.6rem`
- `--mg-badge-font-size`: define badge font size, default: `1.1rem`

<!-- Auto Generated Below -->


## Properties

| Property             | Attribute | Description                                                        | Type               | Default       |
| -------------------- | --------- | ------------------------------------------------------------------ | ------------------ | ------------- |
| `label` _(required)_ | `label`   | Badge label. Include short description. Required for accessibility | `string`           | `undefined`   |
| `outline`            | `outline` | Define if button is using outline style                            | `boolean`          | `undefined`   |
| `value` _(required)_ | `value`   | Badge value                                                        | `number \| string` | `undefined`   |
| `variant`            | `variant` | Define button variant                                              | `string`           | `variants[0]` |


## Dependencies

### Used by

 - [mg-tabs](../../molecules/mg-tabs)

### Graph
```mermaid
graph TD;
  mg-tabs --> mg-badge
  style mg-badge fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
