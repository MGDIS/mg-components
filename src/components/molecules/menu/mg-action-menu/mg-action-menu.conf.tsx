/**
 * List of all possibles placements
 */
export const placements = ['left', 'center', 'right'] as const;

/**
 * Placement type from placements list
 */
export type PlacementType = typeof placements[number];
