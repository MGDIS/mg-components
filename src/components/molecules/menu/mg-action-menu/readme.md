## Use

![](./mg-action-menu/doc/img/mg-action-menu-use.png)

When not all of the items can be displayed because of the width of a [**`<mg-menu> horizontal`**](./?path=/docs/beta-menus-mg-menu--mg-menu) or **`<mg-menu>  mobile`** (TODO) or a [**`<mg-tabs>`**](./?path=/docs/molecules-mg-tabs--mg-tabs), items are replaced by a single "plus" item.

## Behavior

By clicking on the "plus" item a submenu displays the hidden items.

### With an horizontal menu or tabs

![](./mg-action-menu/doc/img/mg-action-menu-behavior-menu-tabs.png)

A **`<mg-menu>  vertical`** displays items in the style of the menu.

Submenu is aligned to the right of the item according to the "submenu of the [**`<mg-menu-item>`**](./?path=/docs/beta-menus-mg-menu-item--mg-menu-item) behavior.

### With a mobile menu

![](./mg-action-menu/doc/img/mg-action-menu-behavior-mobile.png)

A **`<mg-menu> mobile`** displays, over the actual menu, items in the style of the menu.
There are as much lines of items as it is needed to display them.
The **`<mg-action-menu>`** item is replaced by a close button (flat with a cross).

### Badges

![](./mg-action-menu/doc/img/mg-action-menu-mg-menu-large.png)

![](./mg-action-menu/doc/img/mg-action-menu-behavior-badge-mobile.png)

If one or more hidden items have **`<mg-badge>`**, the **`<mg-action-menu>`** item displays a badge with "exclamation" symbol as its content.

## Specs

### Anatomy

Use "ellipsis-vertical" [**`<mg-icon>`**](./?path=/story/atoms-mg-icon--mg-icon) for **`<mg-menu>  horizontal`** or **`<mg-tab>`**.
Use "ellipsis" [**`<mg-icon>`**](./?path=/story/atoms-mg-icon--mg-icon) for **`<mg-menu>  mobile`**.

<!-- Auto Generated Below -->


## Properties

| Property          | Attribute          | Description                            | Type                            | Default         |
| ----------------- | ------------------ | -------------------------------------- | ------------------------------- | --------------- |
| `disableOverflow` | `disable-overflow` | Define component manage child overflow | `boolean`                       | `undefined`     |
| `placement`       | `placement`        | Define the popover menu item size      | `"center" \| "left" \| "right"` | `placements[0]` |


## Dependencies

### Depends on

- [mg-action-more](../mg-action-more)

### Graph
```mermaid
graph TD;
  mg-action-menu --> mg-action-more
  mg-action-more --> mg-popover
  mg-action-more --> mg-button
  mg-action-more --> mg-icon
  mg-action-more --> mg-badge
  mg-action-more --> mg-menu
  mg-action-more --> mg-menu-item
  mg-popover --> mg-button
  mg-popover --> mg-icon
  mg-button --> mg-icon
  mg-menu --> mg-menu-item
  mg-menu --> mg-icon
  mg-menu --> mg-badge
  mg-menu --> mg-menu
  mg-menu-item --> mg-icon
  mg-menu-item --> mg-popover
  style mg-action-menu fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
