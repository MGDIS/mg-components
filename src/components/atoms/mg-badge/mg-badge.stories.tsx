import { h } from '@stencil/core';
import { variants } from './mg-badge.conf';

export default {
  component: 'mg-badge',
  title: 'Atoms/mg-badge',
  argTypes: {
    variant: {
      options: variants,
      control: { type: 'select' },
      table: {
        defaultValue: { summary: variants[0] },
      },
    },
    value: {
      control: { type: 'text' },
    },
  },
};

const Template = args => {
  // Extract slot so it won't be render as an attribute
  // return element
  return <mg-badge {...args}></mg-badge>;
};

export const MgBadge = Template.bind({});
MgBadge.args = {
  value: '99',
  label: 'unread messages',
  variant: variants[0],
  outline: false,
};
