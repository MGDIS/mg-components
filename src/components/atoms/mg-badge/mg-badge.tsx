import { Component, h, Prop, State, Watch } from '@stencil/core';
import { variants, BadgeVariantType } from './mg-badge.conf';
import { ClassList } from '../../../utils/components.utils';
@Component({
  tag: 'mg-badge',
  styleUrl: 'mg-badge.scss',
  shadow: true,
})
export class MgBadge {
  /************
   * Internal *
   ************/

  // Classes
  private classOutline = `mg-badge--outline`;

  /**************
   * Decorators *
   **************/

  /**
   * Badge value
   */
  @Prop({ mutable: true }) value!: string | number;
  @Watch('value')
  validateValue(newValue: MgBadge['value']): void {
    if (`${newValue}`.match(/^(\d+\+*|[?*!a-z])$/i) === null) {
      throw new Error('<mg-badge> prop "value" must be interger or ponctuation character.');
    }
  }

  /**
   * Badge label. Include short description.
   * Required for accessibility
   */
  @Prop() label!: string;
  @Watch('label')
  validateLabel(newValue: MgBadge['label']): void {
    if (typeof newValue !== 'string' || newValue.trim() === '') {
      throw new Error('<mg-badge> prop "label" is required.');
    }
  }

  /**
   * Define button variant
   */
  @Prop() variant?: BadgeVariantType = variants[0]; // info
  @Watch('variant')
  validateVariant(newValue: MgBadge['variant'], oldValue?: MgBadge['variant']): void {
    if (!variants.includes(newValue)) {
      throw new Error(`<mg-badge> prop "variant" must be one of : ${variants.join(', ')}.`);
    } else {
      if (oldValue !== undefined) {
        this.classList.delete(`mg-badge--${oldValue}`);
      }
      this.classList.add(`mg-badge--${newValue}`);
    }
  }

  /**
   * Define if button is using outline style
   */
  @Prop() outline: boolean;
  @Watch('outline')
  validateOutline(newValue: MgBadge['outline']): void {
    if (newValue) this.classList.add(this.classOutline);
    else this.classList.delete(this.classOutline);
  }

  /**
   * Component classes
   */
  @State() classList: ClassList = new ClassList(['mg-badge']);

  /**
   * Check if props are well configured on init
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.validateValue(this.value);
    this.validateLabel(this.label);
    this.validateVariant(this.variant);
    this.validateOutline(this.outline);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <span class={this.classList.join()}>
        {this.value}
        <span class="sr-only">{this.label}</span>
      </span>
    );
  }
}
