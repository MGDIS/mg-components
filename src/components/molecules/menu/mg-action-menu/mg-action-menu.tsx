import { Component, h, Element, Fragment, Prop, Watch } from '@stencil/core';
import { MenuItemSizeType, sizes } from '../mg-menu-item/mg-menu-item.conf';
import { Direction } from '../mg-menu/mg-menu.conf';
import { initLocales } from '../../../../locales';
import { placements, PlacementType, InteractiveElementType, interactivesElements } from './mg-action-menu.conf';
import { OverflowBehavior, OverflowBehaviorElements } from '../../../../utils/behaviors.utils';

type MessageType = { moreLabel: string; badgeLabel: string };

@Component({
  tag: 'mg-action-menu',
  styleUrl: 'mg-action-menu.scss',
  shadow: true,
})
export class MgActionMenu {
  /************
   * Internal *
   ************/

  // constantes
  private readonly mgActionMenu = 'data-mg-action-menu';
  private readonly mgActionMenuInteractive = 'data-mg-action-menu-interactive';
  private readonly mgActionMenuContainer = 'data-mg-action-menu-container';
  private readonly mgActionMenuIcon = 'data-mg-action-menu-icon';
  private readonly mgActionMenuLabel = 'data-mg-action-menu-label';
  private readonly mgActionMenuBadge = 'data-mg-action-menu-badge';

  // variables
  private messages: MessageType;
  private mgActionMenuELement: HTMLElement;
  private mgActionMenuInteractiveElement: HTMLElement;
  private mgActionMenuContainerElement: HTMLElement;
  private overflowBehavior: OverflowBehavior;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgActionMenuElement;

  /**
   * Define the popover menu item size
   */
  @Prop() itemSize: MenuItemSizeType = sizes[0];

  /**
   * Define the popover menu item size
   */
  @Prop({ reflect: true }) placement: PlacementType = placements[0];
  @Watch('placement')
  validatePlacement(newValue: MgActionMenu['placement']): void {
    if (!placements.includes(newValue)) {
      throw new Error(`<mg-action-menu> prop "interactiveElement" must be one off: ${placements.join(',')}.`);
    }
  }

  /**
   * Define interactive element
   * ex: mg-button, mg-menu-item.
   * default: 'mg-button'
   */
  @Prop() interactiveElement: InteractiveElementType = interactivesElements[0];
  @Watch('interactiveElement')
  validateInteractiveElement(newValue: MgActionMenu['interactiveElement']): void {
    if (!interactivesElements.includes(newValue)) {
      throw new Error(`<mg-action-menu> prop "interactiveElement" must be one off: ${interactivesElements.join(',')}.`);
    }
  }

  /************
   * Methods *
   ************/

  /**
   * Render mg-action-menu button element
   * <container>
   *  <item data-mg-action-base-index="0"/>
   *  <item data-mg-action-menu hidden="true|false">
   *   <mg-icon icon="ellipsis-vertical"></mg-icon>
   *   <mg-menu direction="vertical" label="Additional menu">
   *     <item data-mg-action-menu-proxy-index="0" hidden="true|false" />
   *   </mg-menu>
   *  </item>
   * </container>
   *
   * @returns {HTMLElement}
   */
  private renderMgActionMenuELement = (): HTMLElement => {
    // mg-action-menu element
    this.mgActionMenuELement = this.element.querySelector(`[${this.mgActionMenu}]`);
    if (this.mgActionMenuELement === null) {
      this.mgActionMenuELement = document.createElement(this.interactiveElement === 'mg-menu-item' ? 'mg-menu-item' : 'mg-popover');
      this.mgActionMenuELement.setAttribute(this.mgActionMenu, '');
    }
    this.mgActionMenuELement.setAttribute('hidden', '');

    // mg-action-menu interactive element
    if (this.interactiveElement === 'mg-menu-item') {
      this.mgActionMenuELement.setAttribute('size', this.element.querySelector('mg-menu-item').size);
      this.mgActionMenuInteractiveElement = this.mgActionMenuELement;
    } else {
      this.mgActionMenuInteractiveElement = this.element.querySelector(`[${this.mgActionMenuInteractive}]`);
      if (this.mgActionMenuInteractiveElement === null) {
        this.renderInteractiveElement(this.mgActionMenuELement);
      }
    }

    // mg-action-menu container element
    this.mgActionMenuContainerElement = this.element.querySelector(`[${this.mgActionMenuContainer}]`);
    if (this.mgActionMenuContainerElement === null) {
      this.renderContainerElement(this.mgActionMenuELement);
      this.mgActionMenuContainerElement = this.mgActionMenuELement.querySelector(`[${this.mgActionMenuContainer}]`);
    }

    // mg-icon
    if (this.mgActionMenuInteractiveElement.querySelector(`[${this.mgActionMenuIcon}]`) === null) {
      this.renderMgIcon(this.mgActionMenuInteractiveElement);
    }

    // label
    if (this.mgActionMenuInteractiveElement.querySelector(`[${this.mgActionMenuLabel}]`) === null) {
      this.renderMgLabel(this.mgActionMenuInteractiveElement);
    }

    // mg-menu
    if (this.mgActionMenuContainerElement.querySelector('mg-menu') === null) {
      this.renderMgMenu();
    }

    // mg-badge
    if (this.mgActionMenuInteractiveElement.querySelector(`[${this.mgActionMenuBadge}]`) === null) {
      this.renderMgBadge(this.mgActionMenuInteractiveElement, this.mgActionMenuContainerElement.querySelector('mg-menu'));
    }

    this.element.appendChild(this.mgActionMenuELement);

    return this.mgActionMenuELement;
  };

  private isInMenuItem = () => this.interactiveElement === 'mg-menu-item';

  private renderInteractiveElement(mgActionMenuELement: Element) {
    this.mgActionMenuInteractiveElement = document.createElement(this.interactiveElement);
    if (this.interactiveElement === 'mg-button') {
      this.mgActionMenuInteractiveElement.setAttribute('variant', 'secondary');
    }
    mgActionMenuELement.appendChild(this.mgActionMenuInteractiveElement);
  }

  private renderContainerElement(mgActionMenuELement: Element) {
    this.mgActionMenuContainerElement = document.createElement(this.interactiveElement === 'mg-menu-item' ? 'mg-menu' : 'div');
    if (this.interactiveElement === 'mg-menu-item') {
      this.mgActionMenuContainerElement.setAttribute('label', this.messages.moreLabel);
      this.mgActionMenuContainerElement.setAttribute('direction', 'vertical');
    } else {
      this.mgActionMenuContainerElement.setAttribute('slot', 'content');
    }
    this.mgActionMenuContainerElement.setAttribute(this.mgActionMenuContainer, '');
    mgActionMenuELement.appendChild(this.mgActionMenuContainerElement);
  }

  private renderMgBadge(mgActionMenuInteractiveElement: Element, plusMenu: Element) {
    const mgBadge = document.createElement('mg-badge');
    mgBadge.setAttribute(this.mgActionMenuBadge, '');
    mgBadge.setAttribute('label', this.messages.badgeLabel);
    mgBadge.setAttribute('value', '!');
    mgBadge.setAttribute('hidden', '');

    if (this.isInMenuItem()) mgBadge.setAttribute('slot', 'information');

    if (mgActionMenuInteractiveElement) mgActionMenuInteractiveElement.appendChild(mgBadge);
    else this.mgActionMenuELement.insertBefore(mgBadge, plusMenu);
  }

  private renderMgIcon(mgActionMenuInteractiveElement: Element) {
    const mgIcon = document.createElement('mg-icon');
    mgIcon.setAttribute(this.mgActionMenuIcon, '');
    mgIcon.setAttribute('icon', 'ellipsis-vertical');

    if (this.isInMenuItem()) mgIcon.setAttribute('slot', 'image');

    mgActionMenuInteractiveElement.appendChild(mgIcon);
  }

  private renderMgLabel(mgActionMenuInteractiveElement: Element) {
    const label = document.createElement('span');
    label.setAttribute(this.mgActionMenuLabel, '');
    label.classList.add('sr-only');
    label.textContent = this.messages.moreLabel;

    if (this.isInMenuItem()) label.setAttribute('slot', 'label');

    mgActionMenuInteractiveElement.appendChild(label);
  }

  private renderMgMenu() {
    const plusMenu = document.createElement('mg-menu');
    plusMenu.setAttribute('direction', Direction.VERTICAL);
    plusMenu.setAttribute('label', this.messages.moreLabel);
    Array.from(this.element.children).forEach((element, index) => {
      if (element.getAttribute(this.mgActionMenu) === null) {
        element.setAttribute(OverflowBehaviorElements.BASE_INDEX, `${index}`);
        this.renderMgMenuItem(element.textContent, index, plusMenu);
      }
    });

    this.mgActionMenuContainerElement.appendChild(plusMenu);
  }

  private renderMgMenuItem = (label: string, index: number, plusMenu: Element): void => {
    const menuItem = document.createElement('mg-menu-item');
    menuItem.setAttribute('size', this.itemSize);
    menuItem.setAttribute('hidden', '');
    menuItem.setAttribute(OverflowBehaviorElements.PROXY_INDEX, `${index}`);

    const itemLabel = document.createElement('span');
    itemLabel.setAttribute('slot', 'label');
    itemLabel.textContent = label;
    menuItem.appendChild(itemLabel);
    (this.interactiveElement === 'mg-menu-item' ? this.mgActionMenuContainerElement : plusMenu).appendChild(menuItem);

    menuItem.addEventListener('click', event => {
      event.preventDefault();
      this.element.querySelector(`[${OverflowBehaviorElements.PROXY_INDEX}="${index}"]`).dispatchEvent(new CustomEvent('click', { bubbles: true }));
    });
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Validate props
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.messages = initLocales(this.element).messages.plusMenu;
    this.validateInteractiveElement(this.interactiveElement);
    this.validatePlacement(this.placement);
  }

  /**
   * init mg-action-menu
   *
   * @returns {void}
   */
  componentDidLoad(): void {
    this.overflowBehavior = new OverflowBehavior(this.element, this.renderMgActionMenuELement);
    this.overflowBehavior.init();
  }

  disconnectedCallback() {
    this.overflowBehavior.disconnect();
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <Fragment>
        <slot></slot>
      </Fragment>
    );
  }
}
