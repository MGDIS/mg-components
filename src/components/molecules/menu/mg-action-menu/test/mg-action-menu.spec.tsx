import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgActionMenu } from '../mg-action-menu';
import { setupResizeObserverMock } from '../../../../../utils/unit.test.utils';
import { MgBadge } from '../../../../atoms/mg-badge/mg-badge';

const getPage = async args => {
  const page = await newSpecPage({
    components: [MgActionMenu, MgBadge],
    template: () => (
      <mg-action-menu {...args}>
        <ul>
          <li>
            <mg-button>Batman</mg-button>
          </li>
          <li>
            <mg-button>Joker</mg-button>
          </li>
          <li>
            <mg-button>
              Bane
              {args.badgeSlot && <mg-badge value="1" label="movies"></mg-badge>}
            </mg-button>
          </li>
        </ul>
      </mg-action-menu>
    ),
  });

  jest.runAllTimers();
  await page.waitForChanges();

  return page;
};

describe('mg-action-menu', () => {
  let fireMo;
  beforeEach(() => {
    jest.useFakeTimers();
    setupResizeObserverMock({
      observe: function () {
        fireMo = this.cb;
      },
      disconnect: function () {
        return null;
      },
      takeRecords: [],
    });
  });
  afterEach(() => jest.runOnlyPendingTimers());
  describe('render', () => {
    test.each([{}, { badgeSlot: true }])('should manage resize with observer, case %s', async args => {
      const menuSize = 100;
      const page = await getPage(args);
      expect(page.root).toMatchSnapshot();

      const spy = jest.spyOn(page.rootInstance, 'run');
      Object.defineProperty(page.doc.querySelector('ul'), 'offsetWidth', {
        get: jest.fn(() => menuSize),
      });
      Array.from(page.doc.querySelectorAll('li')).forEach(li => {
        Object.defineProperty(li, 'offsetWidth', {
          get: jest.fn(() => 80),
        });
      });

      expect(spy).not.toHaveBeenCalled();

      fireMo([{ contentRect: { width: menuSize }, target: { nodeName: 'UL' } }]);
      await page.waitForChanges();

      expect(spy).toHaveBeenCalledWith(menuSize);
      expect(page.root).toMatchSnapshot();

      // fireMo([{ contentRect: { width: 800 }, target: { nodeName: 'UL' }}]);
      // await page.waitForChanges();

      // expect(spy).toHaveBeenCalledWith(800);
      // expect(page.root).toMatchSnapshot();
    });
  });

  test('should fire disconnect callback', async () => {
    const { root, doc } = await getPage({ label: 'batman menu' });

    expect(root).toMatchSnapshot();

    const mgActionMenu = doc.querySelector('mg-action-menu');

    mgActionMenu.remove();

    expect(root).toMatchSnapshot();
  });
});
