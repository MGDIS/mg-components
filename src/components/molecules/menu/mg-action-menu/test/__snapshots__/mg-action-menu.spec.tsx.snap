// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-action-menu render should manage resize with observer, case { badgeSlot: true } 1`] = `
<mg-action-menu badgeslot="" placement="left">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ul>
    <li>
      <mg-button>
        Batman
      </mg-button>
    </li>
    <li>
      <mg-button>
        Joker
      </mg-button>
    </li>
    <li>
      <mg-button>
        Bane
        <mg-badge>
          <mock:shadow-root>
            <span class="mg-badge mg-badge--info">
              1
              <span class="sr-only">
                movies
              </span>
            </span>
          </mock:shadow-root>
        </mg-badge>
      </mg-button>
    </li>
  </ul>
</mg-action-menu>
`;

exports[`mg-action-menu render should manage resize with observer, case { badgeSlot: true } 2`] = `
<mg-action-menu badgeslot="" placement="left">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ul data-mg-action-menu-base-index="0">
    <li>
      <mg-button>
        Batman
      </mg-button>
    </li>
    <li>
      <mg-button>
        Joker
      </mg-button>
    </li>
    <li>
      <mg-button>
        Bane
        <mg-badge>
          <mock:shadow-root>
            <span class="mg-badge mg-badge--info">
              1
              <span class="sr-only">
                movies
              </span>
            </span>
          </mock:shadow-root>
        </mg-badge>
      </mg-button>
    </li>
  </ul>
  <mg-popover data-mg-action-menu hidden="">
    <mg-button variant="secondary">
      <mg-icon data-mg-action-menu-icon icon="ellipsis-vertical"></mg-icon>
      <span class="sr-only" data-mg-action-menu-label>
        Additional menu
      </span>
      <mg-badge data-mg-action-menu-badge hidden="" label="Include notification(s)" value="!">
        <mock:shadow-root>
          <span class="mg-badge mg-badge--info">
            !
            <span class="sr-only">
              Include notification(s)
            </span>
          </span>
        </mock:shadow-root>
      </mg-badge>
    </mg-button>
    <div data-mg-action-menu-container slot="content">
      <mg-menu direction="vertical" label="Additional menu">
        <mg-menu-item data-mg-action-menu-proxy-index="0" hidden="" size="regular">
          <span slot="label">
            BatmanJokerBane
          </span>
        </mg-menu-item>
      </mg-menu>
    </div>
  </mg-popover>
</mg-action-menu>
`;

exports[`mg-action-menu render should manage resize with observer, case {} 1`] = `
<mg-action-menu placement="left">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ul>
    <li>
      <mg-button>
        Batman
      </mg-button>
    </li>
    <li>
      <mg-button>
        Joker
      </mg-button>
    </li>
    <li>
      <mg-button>
        Bane
      </mg-button>
    </li>
  </ul>
</mg-action-menu>
`;

exports[`mg-action-menu render should manage resize with observer, case {} 2`] = `
<mg-action-menu placement="left">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ul data-mg-action-menu-base-index="0">
    <li>
      <mg-button>
        Batman
      </mg-button>
    </li>
    <li>
      <mg-button>
        Joker
      </mg-button>
    </li>
    <li>
      <mg-button>
        Bane
      </mg-button>
    </li>
  </ul>
  <mg-popover data-mg-action-menu hidden="">
    <mg-button variant="secondary">
      <mg-icon data-mg-action-menu-icon icon="ellipsis-vertical"></mg-icon>
      <span class="sr-only" data-mg-action-menu-label>
        Additional menu
      </span>
      <mg-badge data-mg-action-menu-badge hidden="" label="Include notification(s)" value="!">
        <mock:shadow-root>
          <span class="mg-badge mg-badge--info">
            !
            <span class="sr-only">
              Include notification(s)
            </span>
          </span>
        </mock:shadow-root>
      </mg-badge>
    </mg-button>
    <div data-mg-action-menu-container slot="content">
      <mg-menu direction="vertical" label="Additional menu">
        <mg-menu-item data-mg-action-menu-proxy-index="0" hidden="" size="regular">
          <span slot="label">
            BatmanJokerBane
          </span>
        </mg-menu-item>
      </mg-menu>
    </div>
  </mg-popover>
</mg-action-menu>
`;

exports[`mg-action-menu should fire disconnect callback 1`] = `
<mg-action-menu label="batman menu" placement="left">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ul>
    <li>
      <mg-button>
        Batman
      </mg-button>
    </li>
    <li>
      <mg-button>
        Joker
      </mg-button>
    </li>
    <li>
      <mg-button>
        Bane
      </mg-button>
    </li>
  </ul>
</mg-action-menu>
`;

exports[`mg-action-menu should fire disconnect callback 2`] = `
<mg-action-menu label="batman menu" placement="left">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <ul>
    <li>
      <mg-button>
        Batman
      </mg-button>
    </li>
    <li>
      <mg-button>
        Joker
      </mg-button>
    </li>
    <li>
      <mg-button>
        Bane
      </mg-button>
    </li>
  </ul>
</mg-action-menu>
`;
